bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3ChunkStorage.java

```diff
@@ -0,0 +1,325 @@
+            }
+
+            //Copy the parts
+            for (int i = 0; i < chunks.length; i++) {
+                if (0 != chunks[i].getLength()) {
+                    val sourceHandle = chunks[i];
+                    S3ObjectMetadata metadataResult = client.getObjectMetadata(config.getBucket(),
+                            config.getPrefix() + sourceHandle.getName());
+                    long objectSize = metadataResult.getContentLength(); // in bytes
+                    Preconditions.checkState(objectSize >= chunks[i].getLength());
+                    CopyPartRequest copyRequest = new CopyPartRequest(config.getBucket(),
+                            config.getPrefix() + sourceHandle.getName(),
+                            config.getBucket(),
+                            targetPath,
+                            uploadId,
+                            partNumber++).withSourceRange(Range.fromOffsetLength(0, chunks[i].getLength()));
+
+                    CopyPartResult copyResult = client.copyPart(copyRequest);
+                    partEtags.add(new MultipartPartETag(copyResult.getPartNumber(), copyResult.getETag()));
+                    totalBytesConcatenated += chunks[i].getLength();
+                }
+            }
+
+            //Close the upload
+            client.completeMultipartUpload(new CompleteMultipartUploadRequest(config.getBucket(),
+                    targetPath, uploadId).withParts(partEtags));
+            // Delete all source objects.
+            for (int i = 1; i < chunks.length; i++) {
+                client.deleteObject(config.getBucket(), config.getPrefix() + chunks[i].getName());
+            }
+        } catch (RuntimeException e) {
+            // Make spotbugs happy. Wants us to catch RuntimeException in a separate catch block.
+            // Error message is REC_CATCH_EXCEPTION: Exception is caught when Exception is not thrown
+            throw convertException(chunks[0].getName(), "doConcat", e);
+        } catch (Exception e) {
+            throw convertException(chunks[0].getName(), "doConcat", e);
+        }
+        return totalBytesConcatenated;
+    }
+
+    @Override
+    protected boolean doTruncate(ChunkHandle handle, long offset) throws UnsupportedOperationException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    protected boolean doSetReadOnly(ChunkHandle handle, boolean isReadOnly) throws ChunkStorageException {
+        try {
+            setPermission(handle, isReadOnly ? Permission.READ : Permission.FULL_CONTROL);
+            return true;
+        } catch (Exception e) {
+            throw convertException(handle.getChunkName(), "doSetReadOnly", e);
+        }
+    }
+
+    private void setPermission(ChunkHandle handle, Permission permission) {
+        AccessControlList acl = client.getObjectAcl(config.getBucket(), config.getPrefix() + handle.getChunkName());
```

Comment:
I see that we are concatenating very often `config.getPrefix() + handle.getChunkName()`:
1) it is better to have a common function to compose the final object name
2) we should validate the chunkName and check for invalid values for S3, probably sanitize it by replacing invalid characters
3) what about adding some know separator between the prefix and the chunkName ? this way it won't be possible to mix data from two separate Pravega clusters that share the same S3 bucket. I know this is a remote possibility because "prefix" should be unique and checkName will be some kind of UUID, but adding a check does not cost very much and will make this safer

Reply:
I fixed the first issue. 
Not sure about 2nd and 3rd. These names are generated by us and should not contain invalid chars.